### PR TITLE
[Instrumentation][X86] Limit setting large section flag to medium/large code models

### PR DESCRIPTION
**Pull request by aeubanks**
In #74514 and #74778 we marked various instrumentation-added sections as
large. This causes an extra PT_LOAD segment if using the small code
model. Since people using the small code model presumably aren't hitting
relocation limits, disable this when using the small code model to avoid
the extra segment.

This uses Module::getCodeModel() which isn't necessarily reliable since it reads module metadata (which right now only the clang frontend sets), but it would be nice to get to a point where we reliably put this sort of information (e.g. PIC/code model/etc) in the IR. This requires duplicating the existing tests since opt/llc currently don't set these metadata. If we get to a point where they do set the code model metadata based on command line arguments then we can deduplicate these tests.

### CODE REVIEW

**Comment by aeubanks** [comment]
Actually I'm going to go down the `Module::getCodeModel()` route. Right now I think frontends adding module metadata for the code model is probably not common, but we should aim for that. And this will in the short term still work for Fuchsia without breaking API for other frontends. It just makes updating tests more annoying since now they have to ingest IR with the proper code model metadata.

**Comment by ilovepi** [comment]
> Actually I'm going to go down the `Module::getCodeModel()` route. Right now I think frontends adding module metadata for the code model is probably not common, but we should aim for that. And this will in the short term still work for Fuchsia without breaking API for other frontends. It just makes updating tests more annoying since now they have to ingest IR with the proper code model metadata.

I actually thought it was already available in the Module, and was surprised you had to pass TargetMachine around like that. If I understand where you're going with this, then I think that sounds like a really nice improvement.

**Comment by ilovepi** [comment]
well I guess I should say I thought it would be reliable rather than available, but I see why it wouldn't be.

File: llvm/test/Instrumentation/InstrProfiling/section-code-model-large.ll [hunk 1]
```diff
@@ -0,0 +1,35 @@
+;; Check that certain globals are in large sections under x86-64 large code model (but not in other arches).
+; RUN: opt %s -mtriple=x86_64-unknown-linux -passes=instrprof -S | FileCheck %s
```
Comment (ilovepi) [inline llvm/test/Instrumentation/InstrProfiling/section-code-model-large.ll:2]: The comment makes it sound like this should have a RUN line for other arches. Was that intentional or copy paste from the other tests?

**Review by ilovepi — APPROVED** [review]
LGTM modulo one small comment in the test. Thanks for getting this done so quickly.

**Review by MaskRay — APPROVED** [review]

**Pull request merged**